### PR TITLE
Add HTTP error management

### LINKED ISSUE #1: TODO: Error handling
TODO code:

https://github.com/holaluz/nuxt-clean/blob/master/src/application/Article/createArticle.ts
https://github.com/holaluz/nuxt-clean/blob/master/src/domain/Article/article.ts

is using `throw` the best way to approach error handling in JavaScript?

Interesting read:

https://khalilstemmler.com/articles/enterprise-typescript-nodejs/handling-errors-result-class/

### LINKED ISSUE #5: TODO: finish example
the "Create article" example (from the Vue component to the domain value object) is not functional, and I'd like that very much.

We could even use https://github.com/typicode/json-server to fake a response from the back end.

### PR DESCRIPTION
**Pull request by afontcu**
Fixes #1. Also fixes #5 as the example is somewhat working xD

### CODE REVIEW

File: src/ui/store/article/mutationTypes.ts [hunk 1]
```diff
@@ -1,5 +1,8 @@
 export const ARTICLE = {
-  CREATE_ARTICLE: 'CREATE_ARTICLE',
+  GET_RECENT_ARTICLES_REQUEST: 'GET_RECENT_ARTICLES_REQUEST',
+  GET_RECENT_ARTICLES_SUCCESS: 'GET_RECENT_ARTICLES_SUCCESS',
+  GET_RECENT_ARTICLES_ERROR: 'GET_RECENT_ARTICLES_ERROR',
```
Comment (Samitier) [inline src/ui/store/article/mutationTypes.ts:4]: questions:
 - why not `enum`?
 - why in caps?

File: src/ui/store/article/index.ts [hunk 2]
```diff
@@ -34,37 +34,32 @@ export const getters: GetterTree<RootState, RootState> = {
 }
 
 export const actions: ActionTree<RootState, RootState> = {
-  createArticle({ commit }, editingArticle: Article) {
-    commit(ARTICLE.CREATE_ARTICLE_REQUEST)
+  async getRecentArticles({ commit }) {
```
Comment (Samitier) [inline src/ui/store/article/index.ts:37]: why aren't the name of the actions also an enum?

File: src/ui/store/article/index.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { GetterTree, ActionTree, MutationTree } from 'vuex'
```
Comment (Samitier) [inline src/ui/store/article/index.ts:1]: Pretty much the same as you did but here's how I would do it:

```ts
// actions.ts
import { State } from './state'
export enum Actions {
   foo = 'foo'
}
export const actions: ActionTree<State, State> = {
   [Actions.foo]({ commit }) { ... }
}

// mutations.ts
import { State } from './state'
export enum Mutations {
   foo = 'foo'
}
export const mutations: MutationTree<State> = {
   [Mutation.foo](state, foo:whatever) { ... }
}

// store.ts
export const state = {
  // the object that holds the state
}
export type State = typeof state

//index.ts
export default {
  state() {
    return state
  },
  mutations,
  actions,
}
```

But yeah pretty much the same idea as here

File: src/ui/store/article/mutationTypes.ts (hunk 1 above)
Comment (afontcu) [reply src/ui/store/article/mutationTypes.ts:4]: > why not enum?

because I wrote those before I installed TS 😂  gonna change it up! nice catch

> why in caps?

to be honest, because of convention 🤷 Everywhere I saw people writing mutation types they did so in caps, probably to hint that they are constant values, not something that should be modified.

> why is it called Article and not mutationTypes?

No reason 😇 Probably because I started handling two stores at once in the first example and I wanted to split them. Changing it, too :)

File: src/ui/store/article/index.ts (hunk 3 above)
Comment (afontcu) [reply src/ui/store/article/index.ts:1]: sure! I went ahead and wrote everything down in the same file for brevity, but since this looks promising, I'll go ahead and make it representative of more complex examples

File: src/ui/pages/index.vue [hunk 4]
```diff
@@ -1,55 +1,42 @@
...
-
-  methods: {
-    ...mapActions('article', ['createArticle']),
+  computed: mapState('article', ['loading', 'error', 'articles']),
 
-    handleSubmit(formValues: unknown) {
-      this.createArticle(formValues)
-    },
+  async mounted() {
+    await this.$store.dispatch('article/getRecentArticles')
```
Comment (Samitier) [inline src/ui/pages/index.vue:28]: this is not type safe... and no idea how to do it in vue 2 without decorators :(

maybe we can at least not hardcode the name if we "play" with the `mapActions`?

```ts
methods: {
   ...mapActions(articlesModule, [ArticlesActions.getRecentArticles])
}
```
even though how do we execute the action after? With this abobination? 😂 
```ts
this[ArticlesActions.getRecentArticles]()
```
Or the other is writting a plugin that simply accepts a namespace and an action name as params:
```ts
this.$dispatch(articlesModule, ArticlesActions.getRecentArticles)
```

This for me its pretty important. Kinda makes no sense to type a store if then it's anarchy anyways.

File: src/ui/components/AddArticleForm.vue [hunk 5]
```diff
@@ -1,52 +1,59 @@
...
+            this.message = `I fucked it up: ${e.message}`
+          },
+          respondWithServerError: (e) => {
+            this.message = `They fucked it up: ${e.message}`
+          },
+          respondWithGenericError: (e) => {
+            this.message = `This is fucked up: ${e.message}`
+          },
+        }
+      )
```
Comment (sergigraciaj) [inline src/ui/components/AddArticleForm.vue:56]: I guess this is a method instead of an action just for simplicity, but should the copies/keys for translations live in the VO or if as in this example, are generic errors, live somewhere shared between components?

File: src/ui/pages/index.vue [hunk 6]
```diff
@@ -1,55 +1,42 @@
...
+    ArticleList,
   },
 
-  computed: {
-    ...mapState('article', ['loading', 'error']),
-  },
-
-  methods: {
-    ...mapActions('article', ['createArticle']),
+  computed: mapState('article', ['loading', 'error', 'articles']),
```
Comment (Samitier) [inline src/ui/pages/index.vue:25]: is this type safe too? Does ts infer there are this 3 props in `this`? Why not spreading the `mapState`?

File: src/ui/components/ArticleList.vue [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import { Article } from '@domain/Article'
+
+export default Vue.extend({
+  name: 'ArticleList',
+
+  props: {
+    articles: {
+      type: Array,
+      required: true,
+    } as PropOptions<Article[]>,
```
Comment (Samitier) [inline src/ui/components/ArticleList.vue:22]: Just out of curiosity what happens if you put `Object` instead of `Array` in `type`?

File: src/ui/store/article/index.ts (hunk 2 above)
Comment (afontcu) [reply src/ui/store/article/index.ts:37]: because we cannot benefit from it in our templates when using `mapActions` and so on 😞 I could add a `articlesTypes` and then importing in the component and use `this.$store` somehow, but it looks limited. Also not 100% sure how this would work with namespaced modules 🤔 will take a look!

Lucky us new Vuex releases will come with great TS support, so all these issues will be gone 🎉

File: src/ui/pages/index.vue (hunk 4 above)
Comment (afontcu) [reply src/ui/pages/index.vue:28]: nope, it isn't 😭  this is one of the trade-offs I was talking about when presenting this structure. Vuex is currently not type safe without a lot of gymnastics. There's no way to overcome this limitation due to its current design without introducing a lot of weird stuff such as class decorators and other stuff that might complicate our path to Vuex 4/5 😞 

We can take a look at this, sure! AFAIK, though, there's no "comfortable" solution as of the time of writing

File: src/ui/components/AddArticleForm.vue (hunk 5 above)
Comment (afontcu) [reply src/ui/components/AddArticleForm.vue:56]: Sure, this PR aims to showcase how to deal with HTTP requests and errors, and also that triggering a use case from a component might be a valid option (as long as we isolate our domain layers from the framework). We can explore this further (possibly a use case could return HttpError but also custom Domain Errors), but it is beyond the scope of this PR for now

File: src/ui/components/ArticleList.vue (hunk 7 above)
Comment (afontcu) [reply src/ui/components/ArticleList.vue:22]: Nothing, because the array Array prototype extends from Object 😓 however, the intellisense trusts `Article[]`,  auto-completion provides you with array methods 🎉 

That being said, tagging this prop as `type: Boolean`, for instance, triggers an error:

image

File: src/ui/pages/index.vue (hunk 6 above)
Comment (afontcu) [reply src/ui/pages/index.vue:25]: > is this type safe too?

No :( see image below

> Does ts infer there are this 3 props in this?

Yep! see image below XD

image


> Why not spreading the mapState?

hmm dunno, there's no difference between `computed: mapState('article', ['loading', 'error', 'articles'])` and `computed: { ...mapState('article', ['loading', 'error', 'articles']) },`, right? I guess I was fiddling with this and end up removing the spread xD

File: src/shared/HttpError.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { HttpStatusCode } from './HttpStatusCode'
```
Comment (Samitier) [inline src/shared/HttpError.ts:1]: How do we connect this + typing the responses in axios? Creisi idea: what if we drop axios and just do it with `fetch`?
```ts
export const http = {
   get<T>(url: string): Promise<T> {
      return fetch(url)
        .then(response => {
          if (response.ok) return response.json<T>()
          throw new HttpError(response.statusText)
        })
   }
}
```

Comment (Samitier) [reply src/shared/HttpError.ts:1]: ok I didn't saw httpService and it was literally in front of my face 🤦

Comment (afontcu) [reply src/shared/HttpError.ts:1]: In the error handler from the Http Service:

https://github.com/holaluz/nuxt-clean/pull/7/files#diff-c6a051ea3d1223e9fc3b9a637290275300c2ffba409b47a59d0627eeff5edb35R82-R88

File: src/ui/components/AddArticleForm.vue (hunk 5 above)
Comment (afontcu) [reply src/ui/components/AddArticleForm.vue:56]: Any suggestion (in the form of an issue or a PR) is heavily appreciated, thou!

File: src/infrastructure/Article/ArticleMapper.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+// us with the translation.
+export function fromDTOtoViewModel(dto: ArticleDTO): Article {
+  return {
+    slug: dto.slug,
+    title: dto.title,
+    createdAt: dto.createdAt,
+    body: dto.body,
+    favorited: dto.isFavorited,
+  }
+}
```
Comment (Samitier) [inline src/infrastructure/Article/ArticleMapper.ts:14]: What happens when the DTO directly maps to a domain model? Do we have to write the same thing in both? 

What happens when our domain model is made from multiple dtos?

Do we have to define all the props in a DTO even though we will never use them?

This should throw some kind of error no?

File: src/ui/store/article/index.ts (hunk 2 above)
Comment (afontcu) [reply src/ui/store/article/index.ts]: In the last update I made an enum, even though I'm not 100% sure how to leverage them outside of the store. We might need to iterate on that!

File: src/ui/pages/index.vue (hunk 4 above)
Comment (afontcu) [reply src/ui/pages/index.vue:28]: > Kinda makes no sense to type a store if then it's anarchy anyways

It makes sense ("some" sense, I grant you that!) to have types for everything but parts of the framework, because it helps you think in terms of types, and have type safety when thinking about interactions outside the framework. It'd be great to make the types flow through the whole system, but I do feel there's still a lot of benefit from having them "almost" everywhere!

In short, this is what we have in Vue 2: 
In the store, you have types when interacting with use cases and domain.
In the components, you have types when interacting with data and props.
There's no typing between the store and the components.

File: src/application/Article/getRecentArticles.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+          if (error.isClientError()) {
+            respondWithClientError(error)
+            return
+          }
+
+          respondWithServerError(error)
+        }
+
+        // Example 2.2 - We can also iterate on error constructor
+        if (isHttpError(error)) {
```
Comment (Samitier) [inline src/application/Article/getRecentArticles.ts:62]: I don't think we should check for http errors outside infrastructure. I think we should throw errors that make sense in the context of the useCase from infrastructure. For example:

```ts
// infrastructure/articles.ts
axios.get(...)
.then(mapToArticle)
.catch(e) {
   if (e.status === 404) throw new ArticleDoesNotExist()
   else throw new NetworkError()
}
```

File: src/infrastructure/Article/ArticleMapper.ts (hunk 9 above)
Comment (afontcu) [reply src/infrastructure/Article/ArticleMapper.ts:14]: > What happens when the DTO directly maps to a domain model? Do we have to write the same thing in both?

Hm, I think there's no harm on being explicit, because if the DTO ever changes, we'll have a place to handle that change. I think it is way better than the current solution of just spreading whatever JSON comes from the API, because it leads to API-coupling and also hides the available data. If we think in terms of what events/types we have in our apps, it makes total sense to have translators in the boundaries of our system.

> Do we have to define all the props in a DTO even though we will never use them?

From my POV we should define all the props that make sense in our **domain types**, and use the DTOs to transform that raw information to our types. So I guess the answer is no, there's no need to map 100% of the data a DTO prevents (especially if the endpoint is used in several places, so it serves to several purposes)


> What happens when our domain model is made from multiple dtos?
> This should throw some kind of error no?

io-ts might answer those questions 😄 However, the idea here is to transform **raw data into domain types as soon as possible** so we make sure we're handling the right data inside of our app

File: src/application/Article/getRecentArticles.ts (hunk 10 above)
Comment (afontcu) [reply src/application/Article/getRecentArticles.ts:62]: 100% agree with that. However, 90% of times we'll be handling HTTP errors, so I thought it would make sense to standardize them in some way, to avoid a lot of boilerplate. I agree, though, that it would be much better to return (not throw! 😉) domain-related errors such as `ArticleDoesNotExistError`.

Maybe adopting HTTP Errors as part of the domain for a front-end app makes sense, but I'm totally up to discuss this as it also raised me some concerns 🤗

Comment (afontcu) [reply src/application/Article/getRecentArticles.ts:62]: (btw, I'm using `isHttpError` as a type guard, so that TS understands that whatever Error we're handling inside of the if statement is an Http Error so I can operate on it)

File: src/ui/pages/index.vue (hunk 6 above)
Comment (Samitier) [reply src/ui/pages/index.vue:25]: hahah with having them in `this` it already hyped me <3

File: src/ui/components/ArticleList.vue (hunk 7 above)
Comment (Samitier) [reply src/ui/components/ArticleList.vue:22]: hahah nice!
